File: samples/scala-replicatedentity-examples/src/main/scala/com/example/replicated/multimap/domain/SomeMultiMap.scala
package com.example.replicated.multimap.domain

import kalix.scalasdk.replicatedentity.ReplicatedEntity
import kalix.scalasdk.replicatedentity.ReplicatedEntityContext
import kalix.scalasdk.replicatedentity.ReplicatedMultiMap
import com.example.replicated.multimap
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Kalix tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

class SomeMultiMap(context: ReplicatedEntityContext) extends AbstractSomeMultiMap {

  // tag::update[]
  def put(currentData: ReplicatedMultiMap[String, Double], putValue: multimap.PutValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(currentData.put(putValue.key, putValue.value)) // <1>
      .thenReply(Empty.defaultInstance)

  def putAll(currentData: ReplicatedMultiMap[String, Double], putAllValues: multimap.PutAllValues): ReplicatedEntity.Effect[Empty] =
    effects
      .update(currentData.putAll(putAllValues.key, putAllValues.values)) // <1>
      .thenReply(Empty.defaultInstance)

  def remove(currentData: ReplicatedMultiMap[String, Double], removeValue: multimap.RemoveValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(currentData.remove(removeValue.key, removeValue.value)) // <1>
      .thenReply(Empty.defaultInstance)

  def removeAll(currentData: ReplicatedMultiMap[String, Double], removeAllValues: multimap.RemoveAllValues): ReplicatedEntity.Effect[Empty] =
    effects
      .update(currentData.removeAll(removeAllValues.key)) // <1>
      .thenReply(Empty.defaultInstance)
  // end::update[]

  // tag::get[]
  def get(currentData: ReplicatedMultiMap[String, Double], getValues: multimap.GetValues): ReplicatedEntity.Effect[multimap.CurrentValues] = {
    val values = currentData.get(getValues.key) // <1>
    effects
      .reply(multimap.CurrentValues(getValues.key, values.toSeq))
  }

  /** Command handler for "GetAll". */
  def getAll(currentData: ReplicatedMultiMap[String, Double], getAllValues: multimap.GetAllValues): ReplicatedEntity.Effect[multimap.AllCurrentValues] = {
    val currentValues =
      currentData.keySet.map { key => // <2>
        val values = currentData.get(key)
        multimap.CurrentValues(key, values.toSeq)
      }

    effects.reply(multimap.AllCurrentValues(currentValues.toSeq))
  }
  // end::get[]

}

File: samples/java-replicatedentity-examples/src/main/java/com/example/replicated/multimap/domain/SomeMultiMap.java
package com.example.replicated.multimap.domain;

import kalix.javasdk.replicatedentity.ReplicatedEntityContext;
import kalix.javasdk.replicatedentity.ReplicatedMultiMap;
import com.example.replicated.multimap.SomeMultiMapApi;
import com.google.protobuf.Empty;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class SomeMultiMap extends AbstractSomeMultiMap {
  @SuppressWarnings("unused")
  private final String entityId;

  public SomeMultiMap(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> put(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.PutValue command) {
    return effects()
        .update(multiMap.put(command.getKey(), command.getValue())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> putAll(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.PutAllValues command) {
    return effects()
        .update(multiMap.putAll(command.getKey(), command.getValuesList())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> remove(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.RemoveValue command) {
    return effects()
        .update(multiMap.remove(command.getKey(), command.getValue())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> removeAll(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.RemoveAllValues command) {
    return effects()
        .update(multiMap.removeAll(command.getKey())) // <1>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeMultiMapApi.CurrentValues> get(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.GetValues command) {
    Set<Double> values = multiMap.get(command.getKey()); // <1>
    SomeMultiMapApi.CurrentValues currentValues =
        SomeMultiMapApi.CurrentValues.newBuilder()
            .addAllValues(values.stream().sorted().collect(Collectors.toList()))
            .build();
    return effects().reply(currentValues);
  }

  @Override
  public Effect<SomeMultiMapApi.AllCurrentValues> getAll(
      ReplicatedMultiMap<String, Double> multiMap, SomeMultiMapApi.GetAllValues command) {
    List<SomeMultiMapApi.CurrentValues> allValues =
        multiMap.keySet().stream() // <2>
            .map(
                key -> {
                  List<Double> values =
                      multiMap.get(key).stream().sorted().collect(Collectors.toList());
                  return SomeMultiMapApi.CurrentValues.newBuilder()
                      .setKey(key)
                      .addAllValues(values)
                      .build();
                })
            .collect(Collectors.toList());
    SomeMultiMapApi.AllCurrentValues allCurrentValues =
        SomeMultiMapApi.AllCurrentValues.newBuilder().addAllValues(allValues).build();
    return effects().reply(allCurrentValues);
  }
  // end::get[]
}
